### PR TITLE
Recycle blocks in Flyouts

### PR DESCRIPTION
**Pull request by gnarf**
Re-opening this from #1442 to point at develop for now for better overall PR review / discussion

### Proposed Changes

* Adds the ability for toolbox blocks to be reused between toolbox show calls.  

### Reason for Changes

* Each recyclable block saves us 2-3ms on my macbook pro.

### CODE REVIEW

File: core/block_svg.js [hunk 1]
```diff
@@ -272,27 +272,24 @@ Blockly.BlockSvg.prototype.getIcons = function() {
  * @param {Blockly.BlockSvg} newParent New parent block.
  */
 Blockly.BlockSvg.prototype.setParent = function(newParent) {
-  if (newParent == this.parentBlock_) {
+  var oldParent = this.parentBlock_;
```
Comment (gnarf) [inline core/block_svg.js]: this is a totally separate optimization here (setParent) that will get landed separately - ignore for now, it's just a base difference (i'm on "performance")

File: core/workspace.js [hunk 2]
```diff
@@ -110,6 +110,13 @@ Blockly.Workspace = function(opt_options) {
  */
 Blockly.Workspace.prototype.rendered = false;
 
+/**
```
Comment (gnarf) [inline core/workspace.js]: this also is the diff from another PR (which exists in my "performance" base branch, please ignore for now)

**Review by rachel-fenichel — COMMENTED** [review]
How has this been behaving on the performance branch?

We may want to use the xml representation of the block as the key for recycling (possibly without the ID), rather than just the type.  Then blocks that are being recycled have the same contents at a deep level, instead of just the same type.

The main argument against doing that is if getting the XML representation of a block is expensive.  I haven't benchmarked it.

**Comment by rschamp** [comment]
There are still a few issues with custom procedures and variables on the performance branch which are either caused by this PR or https://github.com/LLK/scratch-gui/pull/1784. We should figure out which.  I think since this PR will make the gains of https://github.com/LLK/scratch-gui/pull/1784 less noticeable, we should try reverting https://github.com/LLK/scratch-gui/pull/1784 out of the performance branch and re-test. What do you think @gnarf?

**Comment by gnarf** [comment]
I think that the GUI PR is the more stable one -- This is probably the culprit -- I'm going to be looking into that today - will update with more info when I have it.

I was gonna do something about it last week but the develop builds were in some weird state that made it hard for me to work on.

**Comment by gnarf** [comment]
so this branch ended up picking up the `block.setParent  during clearing workspace` commit when I rebased from performance and develop --- Just keeping them together

Freshly minted commit 821b13e is "recycle" in its entirety (squashed down to a single commit)

**Comment by gnarf** [comment]
> There are still a few issues with custom procedures

I did make some changes for the `procedures_call` - they will no longer be recycled which seemed to clear up a lot of the smoke test errors.

**Comment by gnarf** [comment]
just rebased on latest merges

**Comment by gnarf** [comment]
Alright - rather than split these back up into two PR's again, could I get a final 👀 from y'all again on landing this?  Remember you can compare these changes (and a few from gui as well) here:

* https://llk.github.io/scratch-gui/performance/#208914317 - includes this branch (flyout-recycle) as well as https://github.com/LLK/scratch-gui/pull/1973
* https://llk.github.io/scratch-gui/#208914317 - "baseline" develop build

File: core/block_svg.js [hunk 3]
```diff
@@ -1385,3 +1393,42 @@ Blockly.BlockSvg.prototype.scheduleSnapAndBump = function() {
...
 };
+
+/**
+ * Determine if this block can be recycled, basic non-dynamic shadows and no variables
+ * @return {boolean} true if the block can be recycled
+ */
+Blockly.BlockSvg.prototype.isRecyclable = function() {
+
+  // The procedures_call type is always dynamic as it can be mutated at runtime.
+  if (this.type === 'procedures_call') {
```
Comment (rachel-fenichel) [inline core/block_svg.js]: What about other things that can be mutated at runtime?  The general rule is probably "if mutationToDom and domToMutation are defined on the block, it's not recyclable".

File: core/procedures.js [hunk 4]
```diff
@@ -237,6 +237,7 @@ Blockly.Procedures.flyoutCategory = function(workspace) {
     //   <mutation ...></mutation>
     // </block>
     var block = goog.dom.createDom('block');
+    block.setAttribute('id', 'proccode:' + mutation.proccode );
```
Comment (rachel-fenichel) [inline core/procedures.js]: Unrelated change?

File: core/block_svg.js (hunk 3 above)
Comment (gnarf) [reply core/block_svg.js]: Thanks! I'm sure this would catch other similar extension blocks in the future. -- Running a quick test to see what else might match this.

File: core/procedures.js (hunk 4 above)
Comment (gnarf) [reply core/procedures.js]: This was one of the first approaches I took, giving the procedure_call's id's - I don't think it hurts anything, but also is probably not necessary - can remove it.

File: core/block_svg.js (hunk 3 above)
Comment (gnarf) [reply core/block_svg.js]: the only other blocks this caught were already `false` by some other check, so I updated accordingly.

**Review by rachel-fenichel — COMMENTED** [review]
Looks good from my side.

File: core/block_svg.js [hunk 5]
```diff
@@ -1385,3 +1393,42 @@ Blockly.BlockSvg.prototype.scheduleSnapAndBump = function() {
...
   }, Blockly.BUMP_DELAY);
 };
+
+/**
+ * Determine if this block can be recycled, basic non-dynamic shadows and no variables
+ * @return {boolean} true if the block can be recycled
+ */
+Blockly.BlockSvg.prototype.isRecyclable = function() {
+
+  // If the block needs to parse mutations, it's probably safest to never recycle.
```
Comment (kchadha) [inline core/block_svg.js:1403]: +1

**Review by rschamp — APPROVED** [review]
Thank you @gnarf!
